File: src/RestSharp.Portable.Hal/RestSharp.Portable.Hal.Tests/HalClientTests.fs
module RestSharp.Portable.Hal.Tests

open HalClientTestInfra
open NUnit.Framework
open FsUnit
open System.Net
open Newtonsoft.Json.Linq
open RestSharp.Portable.Hal.Helpers

type HalTests() = 

    let mutable client:HalClient = TestConfig.CreateClient()

    [<SetUp>]
    member test.Setup () =
        client <- TestConfig.CreateClient() 
        ()

    [<Test>]
    member test.``should be able to get resource`` () =
        let resource = client.From("api/cardholders").GetAsync() |> Async.RunSynchronously
        resource |> should be ofExactType<Resource>

    [<Test>]
    member test.``should be able to get resource following link`` () =
        let resource = 
            client.From("api/cardholders")
                .Follow("register")
                .GetAsync() |> Async.RunSynchronously
        
        let tjo = resource.Parse<'RegisterForm>()
        tjo.id === -1
    
    [<Test>]
    member test.``should be able to get resource following multiple links`` () =
        let resource = 
            client.From("api/cardholders")
                .Follow("register")
                .Follow("self")
                .GetAsync<RegistrationForm>() |> Async.RunSynchronously
        
        resource.id === -1
    
    [<Test>]
    member test.``should follow templated link`` () = 
        let resource = 
            client.From("api/cardholders")
                .Follow("cardholder", ["id" => "123"])
                .GetAsync<CardHolderDetails>() |> Async.RunSynchronously
        
        let expected = {
            CardHolderDetails.id = 123;
            name = "Customer Number123";
            anotherCard = { idAgain = "again" }
        }

        resource === expected
   
    [<Test>]
    member test.``should allow url segment state`` () = 
        let resource = 
            client.From("api/cardholders")
                .UrlSegments(["id" => "123"])
                .Follow("cardholder")
                .GetAsync<CardHolderDetails>() |> Async.RunSynchronously
        
        let expected = {
            CardHolderDetails.id = 123;
            name = "Customer Number123";
            anotherCard = { idAgain = "again" }
        }

        resource === expected
        
    [<Test>]
    member test.``provided url segment should take precedence`` () = 
        let resource = 
            client.From("api/cardholders")
                .UrlSegments(["id" => "123"])
                .Follow("cardholder", ["id" => "112"])
                .GetAsync<CardHolderDetails>() |> Async.RunSynchronously
        
        let expected = {
            CardHolderDetails.id = 112;
            name = "Customer Number112";
            anotherCard = { idAgain = "again" }
        }

        resource === expected

    [<Test>]
    member test.``json nuances`` () = 
        let resource = 
            client.From("api/cardholders")
                //.Follow("cardholder", [("id", "112")])
                .Follow("cardholder", ["id" => "112"])
                .GetAsync()
                |> Async.RunSynchronously
        
        let embedded = resource.data.["_embedded"]
        embedded |> should not' (be Null)

        let target = embedded.["card"]
        target |> should not' (be Null)

        let ne = Newtonsoft.Json.Linq.JObject.Parse("{foo:20}")
        let wn = Newtonsoft.Json.Linq.JObject.Parse("{foo:20, _embedded:null}") //pain
        let wo = Newtonsoft.Json.Linq.JObject.Parse("{foo:20, _embedded:22}")
        let we = Newtonsoft.Json.Linq.JObject.Parse("{foo:20, _embedded:{card:1234}}")
        let wen = Newtonsoft.Json.Linq.JObject.Parse("{foo:20, _embedded:{card:null}}") //pain
        let wenc = Newtonsoft.Json.Linq.JObject.Parse("{foo:20, _embedded:{foo:null}}")
        
        ne.["_embedded"] |> should be Null
        wo.["_embedded"] |> should not' (be Null)
        we.["_embedded"].["card"] |> should not' (be Null)
        wenc.["_embedded"].["card"] |> should be Null

        wn.["_embedded"].Type === Newtonsoft.Json.Linq.JTokenType.Null


    [<Test>]
    member test.``should follow multiple rels in one go`` () = 
        let resource = 
            client.From("api/cardholders")
                .UrlSegments(["id" => "112"])
                .Follow(["cardholder"; "card"])
                .GetAsync<CardEmbedded>() |> Async.RunSynchronously
        
        let expected = { CardEmbedded.number = 101; ``type``="mastercard" }

        resource === expected

    [<Test>]
    member test.``merging forms`` () = 
        let data = { LoadCardForm.amount = 100M; currency="GBP" }
        let resource = JObject.Parse("{amount:0, currency:'USD', unknown:22, _links: {self:'/api/foo'}, _embedded:{card:{}}}")

        let merged = merge resource data
       
        merged.Value<int>("amount") === 100
        merged.Value<string>("currency") === "GBP"
        merged.Value<int>("unknown") === 22
        merged.["_links"] |> should be Null
        merged.["_embedded"] |> should be Null

    [<Test>]
    member test.``should follow location header`` () =
        let newData = {RegistrationForm.id = 55; name="Johny"}
        let resource = 
            async{
                let! form = client.From("api/cardholders")
                                .Follow("register").GetAsync()
                return! form.PostAsync(newData)             
            } |> Async.RunSynchronously

        let nr = 
            resource.FollowHeader("Location")
                .GetAsync<CardHolderDetails>() |> Async.RunSynchronously
        
        nr.name === "Customer Number55" //irl would be Johny but server does not persist
        nr.anotherCard.idAgain === "again"

    [<Test>]
    member test.``should follow location header and continue traversal`` () =
        let newData = {RegistrationForm.id = 55; name="Johny"}
        let nr = 
            async{
                let! form = 
                    client.From("api/cardholders")
                        .Follow("register").GetAsync()
                let! resource = form.PostAsync(newData)

                return!
                    resource.FollowHeader("Location")
                        .Follow("updatecardholder")
                        .GetAsync<UpdateCardHolderForm>()
            } |> Async.RunSynchronously

        nr.id === 0

    [<Test>]
    member test.``should follow from resource``() = 
        //i.e. from resource response, should be able to continue traversal
        let resource = 
            client.From("api/cardholders")
                .GetAsync() |> Async.RunSynchronously
        let next = resource.Follow("register")
                    .GetAsync<RegistrationForm>() |> Async.RunSynchronously
        
        next.id === -1

    [<Test>]
    member test.``should follow with segments from resource``() = 
        //i.e. from resource response, should be able to continue traversal
        let resource = 
            client.From("api/cardholders").GetAsync() |> Async.RunSynchronously
        let next =
                resource
                    .Follow("cardholder", ["id" => "123"])
                    .GetAsync<CardHolderDetails>() |> Async.RunSynchronously
        
        let expected = {
            CardHolderDetails.id = 123;
            name = "Customer Number123";
            anotherCard = { idAgain = "again" }
        }

        next === expected     
        
    [<Test>]
    member test.``should follow multiple times with segments from resource``() = 
        //i.e. from resource response, should be able to continue traversal
        let resource = 
            client.From("api/cardholders").GetAsync() |> Async.RunSynchronously
        let next =
                resource
                    .Follow("cardholder", ["id" => "123"])
                    .GetAsync() |> Async.RunSynchronously
        let next2 = next.Follow("self").GetAsync<CardHolderDetails>() |> Async.RunSynchronously
        
        let expected = {
            CardHolderDetails.id = 123;
            name = "Customer Number123";
            anotherCard = { idAgain = "again" }
        }

        next2 === expected         

    [<Test>]
    member test.``should be able to post from fetched resource``() = 
        let newData = {RegistrationForm.id = 55; name="Johny"}
        let form = 
            client.From("api/cardholders")
                .Follow("register").GetAsync() |> Async.RunSynchronously
        
        let resource = form.PostAsync(newData) |> Async.RunSynchronously
       
        let locationHeader = resource.response.Headers.GetValues("Location") |> Seq.head

        resource.response.StatusCode === HttpStatusCode.Created
        locationHeader === "/api/cardholders/55" 

    [<Test>]
    member test.``should be able to post from fetched resource (async example)``() = 
        let newData = {RegistrationForm.id = 55; name="Johny"}
        
        let finalResult = 
            async{
                let! form = 
                    client.From("api/cardholders")
                             .Follow("register").GetAsync() 
                //let formData = form.Parse<>() # use parsed data

                let! postResult = form.PostAsync(newData)
                return postResult
            } |> Async.RunSynchronously

        let locationHeader = finalResult.response.Headers.GetValues "Location" |> Seq.head
        
        locationHeader === "/api/cardholders/55"
        finalResult.response.StatusCode === HttpStatusCode.Created 
    
    [<Test>]
    member test.``should put to server from resource`` () =
        let newData = {RegistrationForm.id = 55; name="Johny"}
        let resource = 
            async {
                let! form = 
                    client.From("api/cardholders")
                        .Follow("register").GetAsync()
                return! form.PutAsync(newData) 
            } |> Async.RunSynchronously

        let locationHeader = resource.response.Headers.GetValues("Location") |> Seq.head

        resource.response.StatusCode === HttpStatusCode.Created
        locationHeader === "/api/cardholders/55"

    [<Test>]
    member test.``should delete to server from resource`` () =
        let newData = {RegistrationForm.id = 55; name="Johny"}
        let resource =
            async{ 
                let! form = client.From("api/cardholders")
                                .Follow("register").GetAsync()
                return! form.DeleteAsync(newData) }|> Async.RunSynchronously
        
        let locationHeader = resource.response.Headers.GetValues("Location") |> Seq.head

        resource.response.StatusCode === HttpStatusCode.OK
        locationHeader === "/api/cardholders" 

File: src/RestSharp.Portable.Hal/RestSharp.Portable.CSharpTests/HalClientTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Http;
using Hal;
using Microsoft.Owin.Testing;
using NUnit.Framework;
using RestSharp.Portable.CSharpTests;
using RestSharp.Portable.Hal.CSharp;
using RestSharp.Portable.HttpClientImpl;
using Simple.Data.Extensions;

namespace RestSharp.Portable.CSharpTests
{
    public class TestHttpClientFactory : DefaultHttpClientFactory
    {
        private readonly TestServer _server;
        private readonly DelegatingHandler[] _delegatingHandlers;

        public TestHttpClientFactory(TestServer server, params DelegatingHandler[] delegatingHandlers)
        {
            _server = server;
            _delegatingHandlers = delegatingHandlers;
        }

        public override HttpClient CreateClient(IRestClient client, IRestRequest request)
        {
            var handler =  HttpClientFactory.CreatePipeline(_server.Handler, _delegatingHandlers);
            return new HttpClient(handler){BaseAddress = new Uri("http://localhost")};
        }

    }
    
    [TestFixture]
    public class HalClientTests
    {

        private HalClient _client;
        private TestServer _server;

        [TestFixtureSetUp]
        public void FixtueSetup()
        {
            _server = TestServer.Create<Startup>();
        }

        [SetUp]
        public void SetUp()
        {
            var clientFactory = new HalClientFactory()
                 .Accept("application/hal+json")
                 .HttpClientFactory(new TestHttpClientFactory(_server));
            _client = clientFactory.CreateHalClient("http://dummy-unsused");
        }

        [Test]
        public void should_get_resource()
        {
            var resource = _client.From("api/cardholders").GetAsync().Result;
            Assert.IsInstanceOf<Resource>(resource);
        }

        [Test]
        public void should_get_resource_following_link()
        {
            var resource = _client.From("api/cardholders")
                .Follow("register")
                .GetAsync().Result;

            var tjo = resource.Parse<RegistrationForm>();

            Assert.AreEqual(-1, tjo.Id);
        }

        [Test]
        public void should_get_resource_following_multiple_links()
        {
            var resource = _client.From("api/cardholders")
                .Follow("register")
                .Follow("self")
                .GetAsync().Result;

            var tjo = resource.Parse<RegistrationForm>();

            Assert.AreEqual(-1, tjo.Id);
        }

        [Test]
        public void should_follow_templated_link()
        {
            var resource = _client.From("api/cardholders")
                .Follow("cardholder", new { id = "123" })
                .GetAsync<CardHolderDetails>().Result;

            Assert.AreEqual(123, resource.Id);
            Assert.AreEqual("Customer Number123", resource.Name);
            Assert.AreEqual("again", resource.AnotherCard.IdAgain);
        }

        [Test]
        public void should_allow_url_segment_state()
        {
            var resource = _client.From("/api/cardholders")
                .UrlSegments(new { id = "123" })
                .Follow("cardholder")
                .GetAsync<CardHolderDetails>().Result;

            Assert.AreEqual(123, resource.Id);
            Assert.AreEqual("Customer Number123", resource.Name);
            Assert.AreEqual("again", resource.AnotherCard.IdAgain);
        }


        [Test]
        public void provided_url_segment_should_take_precedence()
        {
            var resource = _client.From("/api/cardholders")
                .UrlSegments(new { id = "123" })
                .Follow("cardholder", new{id="112"})
                .GetAsync<CardHolderDetails>().Result;

            Assert.AreEqual(112, resource.Id);
            Assert.AreEqual("Customer Number112", resource.Name);
            Assert.AreEqual("again", resource.AnotherCard.IdAgain);
        }

        [Test]
        public void should_follow_multiple_rels_in_one_go()
        {
            var resource = _client.From("/api/cardholders")
                .UrlSegments(new { id = "112" })
                // ReSharper disable once PossiblyMistakenUseOfParamsMethod
                .Follow("cardholder", "card")
                .GetAsync<CardEmbedded>().Result;

            Assert.AreEqual("101", resource.Number);
            Assert.AreEqual("mastercard", resource.Type);
        }

        [Test]
        public void should_get_embedded_resource()
        {
            var resource = _client.From("/api/cardholders")
                // ReSharper disable once PossiblyMistakenUseOfParamsMethod
                .Follow("cardholder", new{id=112})
                .Follow("card")
                .GetAsync<CardEmbedded>().Result;

            Assert.AreEqual("101", resource.Number);
            Assert.AreEqual("mastercard", resource.Type);
        }

        [Test]
        public void should_follow_link_in_embedded_resource()
        {
            var resource = _client.From("/api/cardholders")
                .Follow("cardholder", new { id = 112 })
                .Follow("card")
                .Follow("loadcard")
                .GetAsync<LoadCardForm>()
                .Result;

            Assert.AreEqual(100M, resource.Amount);
            Assert.AreEqual("GBP", resource.Currency);
        }

        [Test]
        public void should_post_form_to_server()
        {
            var newData = new RegistrationForm {Id = 55, Name = "Johny"};
            var resource =
                _client.From("/api/cardholders")
                    .Follow("register")
                    .PostAsync(newData)
                    .Result;

            Assert.AreEqual(HttpStatusCode.Created, resource.Response.StatusCode);

            var location = resource.Response.Headers.GetValues("Location").First();
            Assert.AreEqual("/api/cardholders/55", location);
        }

        [Test]
        public void should_post_form_to_server_and_parse_body_if_you_want()
        {
            var newData = new RegistrationForm { Id = 55, Name = "Johny" };
            var resource =
                _client.From("/api/cardholders")
                    .Follow("register")
                    .PostAsync(newData)
                    .Result;

            var body = resource.Parse<CardHolderDetails>();

            Assert.AreEqual("Johny", body.Name);
            Assert.AreEqual("lala", body.AnotherCard.IdAgain);
        }

        [Test]
        public void should_post_form_to_server_and_parse_body_nicely_if_you_want()
        {
            var newData = new RegistrationForm { Id = 55, Name = "Johny" };
            var resource =
                _client.From("/api/cardholders")
                    .Follow("register")
                    .PostAsyncAndParse<CardHolderDetails>(newData)
                    .Result;

            Assert.AreEqual("Johny", resource.Name);
            Assert.AreEqual("lala", resource.AnotherCard.IdAgain);
        }

        [Test]
        public void should_put_to_server()
        {
            var newData = new RegistrationForm { Id = 55, Name = "Johny" };
            var resource =
                _client.From("/api/cardholders")
                    .Follow("register")
                    .PutAsync(newData)
                    .Result;

            Assert.AreEqual(HttpStatusCode.Created, resource.Response.StatusCode);

            var location = resource.Response.Headers.GetValues("Location").First();
            Assert.AreEqual("/api/cardholders/55", location);
        }

        [Test]
        public void should_delete_to_server()
        {
            var newData = new RegistrationForm { Id = 55, Name = "Johny" };
            var resource =
                _client.From("/api/cardholders")
                    .Follow("register")
                    .DeleteAsync(newData)
                    .Result;

            Assert.AreEqual(HttpStatusCode.OK, resource.Response.StatusCode);

            var location = resource.Response.Headers.GetValues("Location").First();
            Assert.AreEqual("/api/cardholders", location);
        }

        [Test]
        public void should_follow_location_header()
        {
            var newData = new RegistrationForm { Id = 55, Name = "Johny" };
            var resource =
                _client.From("/api/cardholders")
                    .Follow("register")
                    .PostAsync(newData)
                    .Result;

            var newResource = resource.FollowHeader("Location").GetAsync<CardHolderDetails>().Result;

            Assert.AreEqual("Customer Number55", newResource.Name);
            Assert.AreEqual("again", newResource.AnotherCard.IdAgain);
        }

        [Test]
        public void should_follow_location_header_and_continue_traversal()
        {
            var newData = new RegistrationForm { Id = 55, Name = "Johny" };
            var resource =
                _client.From("/api/cardholders")
                    .Follow("register")
                    .PostAsync(newData)
                    .Result;

            var newResource = resource.FollowHeader("Location")
                .Follow("updatecardholder")
                .GetAsync<UpdateCardHolderForm>()
                .Result;

            Assert.AreEqual(0, newResource.Id);
        }

        [Test]
        public void can_pass_in_custom_httpClientFactory()
        {
            Assert.IsNotNull(_client);
        }

        
    }
}